src/main/java/build/buildfarm/instance/memory/MemoryInstance.java

```diff
@@ -207,6 +208,19 @@ protected void onQueue(Operation operation, Action action) {
     putBlob(action.toByteString());
   }
 
+  @Override
+  protected void validateAction(Action action) {
+    if (action.hasTimeout() && config.hasMaximumActionTimeout()) {
+      Duration timeout = action.getTimeout();
+      Duration maximum = config.getMaximumActionTimeout();
+      Preconditions.checkState(
```

Comment:
Maybe not in this patch, but these precondition violations are of the kind that you want to propagate back to the user as a FAILED_PRECONDITION status, with proper details, so that the user sees the error in Bazel and knows what do they need to fix. One way to do it would be to throw a proper StatusRuntimeException from here, and let your new interceptor handle it. I do it almost right in https://github.com/bazelbuild/bazel/blob/master/src/tools/remote/src/main/java/com/google/devtools/build/remote/worker/StatusUtils.java (some of my status codes are wrong, for example it should be FAILED_PRECONDITION instead of NOT_FOUND for missing inputs/command https://github.com/googleapis/googleapis/blob/master/google/devtools/remoteexecution/v1test/remote_execution.proto#L87), but at least it allows to distinguish between user-actionable errors and INTERNAL_ERROR, retriable vs non-retriable, etc.

Reply:
Agreed, I have another diff in the pipeline that will catch IllegalStateException thrown by Preconditions in validateAction (which will declare the throw), and translate in either Instance::execute() or ExecutionService::execute() into StatusRuntimeException (depending on whether I care about introducing grpc deps at the AbstractInstanceServer level). Might just end up translating to differentiate IllegalState in validation from IllegalState during subsequent internal execute operations

Reply:
Updated #46 with the change to respond with FAILED_PRECONDITION on Precondition violations.